sdk/helper/storagepacker/storagepacker_v2.go

```diff
@@ -81,6 +81,156 @@ type LockedBucket struct {
 	*Bucket
 }
 
+type lockOperation func(*LockedBucket)
+
+// Safely identify the bucket for a given key, and aquire its storage lock
+// The key may be the full hash of the item ID, or a prefix, with a bucket Key with embedded /'s
+// The cache may initially be empty, so "not found" is acceptable, and needs to be handled
+// by deciding whether to create the bucket.
+//
+// This function uses the same logic whether we want a read or a write operation.
+// Use the helpers below that set everything up nicely for read locks or write locks.
+func (s *StoragePackerV2) lockBucket(key string, acquire lockOperation, release lockOperation) (bucket *LockedBucket, found bool, err error) {
+	cacheKey := s.GetCacheKey(key)
+	lastCacheKey := ""
+
+	for true {
+		s.bucketsCacheLock.RLock()
+		keyPrefix, bucketRaw, found := s.bucketsCache.LongestPrefix(cacheKey)
+		s.bucketsCacheLock.RUnlock()
+
+		if !found {
+			return nil, false, nil
+		}
+
+		bucket := bucketRaw.(*LockedBucket)
+		acquire(bucket)
+		if !bucket.HasShards {
+			// Found a leaf bucket.
+			// Lock still held on return
+			return bucket, true, nil
+		} else if keyPrefix != lastCacheKey {
+			// Try again, we moved down the tree
+			lastCacheKey = keyPrefix
+			release(bucket)
+		} else {
+			finalKey := bucket.Key
+			release(bucket)
+			return nil, true, fmt.Errorf("Bucket %s has shards but no longer prefix found.",
+				finalKey)
+		}
+	}
+	// Not possible
+	return nil, false, nil
+}
+
+func readAcquire(b *LockedBucket) {
+	b.RLock()
+}
+func readRelease(b *LockedBucket) {
+	b.RUnlock()
+}
+func writeAcquire(b *LockedBucket) {
+	b.Lock()
+}
+func writeRelease(b *LockedBucket) {
+	b.Unlock()
+}
+
+func (s *StoragePackerV2) lockBucketForRead(key string) (bucket *LockedBucket, found bool, err error) {
+	return s.lockBucket(key, readAcquire, readRelease)
+}
+
+func (s *StoragePackerV2) lockBucketForWrite(key string) (bucket *LockedBucket, found bool, err error) {
+	return s.lockBucket(key, writeAcquire, writeRelease)
+}
+
+// Return the topmost bucket in the tree for a given key.
+// Used as a defult if the cache is empty or bypassed.
+func (s *StoragePackerV2) firstKey(key string) (string, error) {
+	cacheKey := s.GetCacheKey(key)
+	rootShardLength := s.BaseBucketBits / 4
+	if len(cacheKey) < rootShardLength {
+		return cacheKey, errors.New("Key too short.")
+	}
+	return cacheKey[0 : s.BaseBucketBits/4], nil
+}
+
+// If an entry is not found in bucketsCache, it might be because it's never
+// been created. Try to load it from storage first, then create it if desired.
```

Comment:
Why would we want to create it here?

Reply:
Sure, I thought there might be more than one place where we want to create or load it, but that turned out not to be the case.

However, the GetBucket() interface has an issue that I haven't addressed which is relevant.  If you GetBucket() and there's no existing bucket, we return nil.  If the caller wanted to create the bucket they would have to reach deep into the implementation to acquire the correct bucket lock, because we didn't return one to them. 

But GetBucket() has other problems, the caller has to be aware that the bucket may have been sharded between the time it was returned to them, and the time they managed to (re-)acquire the bucket lock.

Reply:
I'd really prefer to eliminate the "load" branch on the assumption that we always go through the preload.  But that wasn't the assumption that was in place when I started work.

Reply:
I'm not sure of the existing usages of GetBucket and PutBucket external to the library, but this reminds me of the conversation we had about making at least some of those unexported and instead allowing variadic args in PutItem, which I think would solve this issue? I wonder if there is really a reason to ever allow direct bucket access (outside of legacy code for spv1 which will be converting to the new interfaces).